src/main/java/dev/openfeature/sdk/OpenFeatureAPI.java

```diff
@@ -135,19 +185,123 @@ public List<Hook> getHooks() {
      * {@inheritDoc}
      */
     public void clearHooks() {
-        try (AutoCloseableLock __ = hooksLock.writeLockAutoCloseable()) {
+        try (AutoCloseableLock __ = lock.writeLockAutoCloseable()) {
             this.apiHooks.clear();
         }
     }
 
     public void shutdown() {
         providerRepository.shutdown();
+        // TODO: shutdown events
     }
 
     /**
-     * This method is only here for testing as otherwise all tests after the API shutdown test would fail.
+     * {@inheritDoc}
+     */
+    @Override
+    public OpenFeatureAPI onProviderReady(Consumer<EventDetails> handler) {
+        return this.on(ProviderEvent.PROVIDER_READY, handler);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public OpenFeatureAPI onProviderConfigurationChanged(Consumer<EventDetails> handler) {
+        return this.on(ProviderEvent.PROVIDER_CONFIGURATION_CHANGED, handler);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public OpenFeatureAPI onProviderStale(Consumer<EventDetails> handler) {
+        return this.on(ProviderEvent.PROVIDER_STALE, handler);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public OpenFeatureAPI onProviderError(Consumer<EventDetails> handler) {
+        return this.on(ProviderEvent.PROVIDER_ERROR, handler);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public OpenFeatureAPI on(ProviderEvent event, Consumer<EventDetails> handler) {
+        try (AutoCloseableLock __ = lock.writeLockAutoCloseable()) {
+            this.eventSupport.addGlobalHandler(event, handler);
+            return this;
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public OpenFeatureAPI removeHandler(ProviderEvent event, Consumer<EventDetails> handler) {
+        this.eventSupport.removeGlobalHandler(event, handler);
+        return this;
+    }
+
+    void removeHandler(String clientName, ProviderEvent event, Consumer<EventDetails> handler) {
+        try (AutoCloseableLock __ = lock.writeLockAutoCloseable()) {
+            eventSupport.removeClientHandler(clientName, event, handler);
+        }
+    }
+
+    void addHandler(String clientName, ProviderEvent event, Consumer<EventDetails> handler) {
+        try (AutoCloseableLock __ = lock.writeLockAutoCloseable()) {
+            // if the provider is READY, run immediately
+            if (event.equals(ProviderEvent.PROVIDER_READY)
+                    && ProviderState.READY.equals(this.providerRepository.getProvider(clientName).getState())) {
+                eventSupport.runHandler(handler, EventDetails.builder().clientName(clientName).build());
```

Comment:
I think we should run the handler outside of the write lock

Reply:
I think, in effect, we are.

The handler is run in a new thread by the executor. This thread doesn't block when calling `eventSupport.runHandler`.

Reply:
Gotcha, thanks for the explanation 👍